connectors/crowdstrike/connector/server_access.py

```diff
@@ -59,13 +63,37 @@ def get_hosts(self, last_model_state_id=None):
                     error_detail = host_lookup["body"]["errors"]
                     for err in error_detail:
                         raise DatasourceFailure(err["message"], err["code"])
-                if len(hosts) < host_lookup['body']['meta']['pagination']['total']:
-                    offset = len(hosts)
-                    host_lookup = discover.query_hosts(filter=result_filter, limit=DISCOVER_PAGE_SIZE, offset=offset)
+
+                if len(host_detail["body"]["resources"]) == DISCOVER_PAGE_SIZE:
+                    '''CrowdStrike's Discover endpoint has the following limitation: limit + offset <= 10,000
+                        For that reason we need to do additional queries to get all hosts. We do this by making
+                        additional queries and filter on the last_seen_timestamp. This also means we have the
+                        potential of collecting duplicate hosts. So we make sure to remove these duplicates.'''
+                    if DISCOVER_PAGE_SIZE + offset > 10000:
+                        offset = 0
+                        last_added = deep_get(hosts[-1], ['last_seen_timestamp'])
+                        last_added_filter = "last_seen_timestamp:>='" + last_added + "'"
+                        if result_filter == "":
+                            updated_filter = last_added_filter
+                        else:
+                            updated_filter = result_filter + " + " + last_added_filter
+
+                        host_lookup = discover.query_hosts(filter=updated_filter, limit=DISCOVER_PAGE_SIZE, sort=sort)
+                    else:
+                        offset = len(hosts)
+                        host_lookup = discover.query_hosts(filter=result_filter, limit=DISCOVER_PAGE_SIZE, offset=offset, sort=sort)
                 else:
                     is_paged_result = False
+
+            if len(hosts) > total:
+                # Need to remove potential duplicates due to the additional grouping we do if the total number of hosts is >10,000
+                context().logger.info("Removing duplicate hosts added during import.")
+                hosts = remove_duplicates(hosts)
+            
         except Exception as ex:
-            raise DatasourceFailure(ex)
+            raise DatasourceFailure(ex)      
+
+        context().logger.info("Imported %s hosts from the CrowdStrike Falcon environment.", len(hosts))
         return hosts
 
     def get_applications(self, app_filter=None, last_model_state_id=None):
```

Comment:
For the remaining rest calls can you do something similar for debugging. This way we can keep track of what calls are being made and what the parameters of those calls are?